Refactor `app/gui/queryresults/queryresultsservice.js`

Raruto opened the pull request:
## Depends on: https://github.com/g3w-suite/g3w-client/pull/452

## List of changes:

- add tons `@FIXME` comments (eg. please `"add function description"` here)
- code cleanup (eg. tons of weird [ternary conditions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator), [short-circuiting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND#short-circuit_evaluation) logic and missing [brackets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/if...else)!)
- full function rewrite `QueryResultsService::updateLayerResultFeatures(layer)`
- notable function rewrite `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`
- notable function rewrite `QueryResultsService::_addRemoveSelectionFeature(layer, feature, index, force)`
- notable function rewrite `QueryResultsService::setActionsForLayers(layers, options = { add: false })`
- removed function `QueryResultsService::_digestFeaturesForLayers(featuresForLayers)` (replaced by `QueryResultsService::setQueryResponse`)
- deprecate function `QueryResultsService.prototype.downloadApplicationWrapper(downloadFnc, options)` in favor of `GUI::downloadWrapper(downloadFnc, options)`

Closes: https://github.com/g3w-suite/g3w-client/issues/58
Closes: https://github.com/g3w-suite/g3w-client/issues/59

Preparatory for: https://github.com/g3w-suite/g3w-client/issues/84
Related to: https://github.com/g3w-suite/g3w-client/issues/193

---

~~**NB** this pull depends on https://github.com/g3w-suite/g3w-client/pull/308 for the addition of the [`"util-deprecate"`](https://www.npmjs.com/package/util-deprecate) dependency (so it is recommended to merge the other one first to keep things tidy)~~

Raruto committed:
refactor `src/app/gui/queryresults/queryresultsservice.js`

- add FIXME comments (eg. please "add function description")
- code cleanup (eg. tons of weird ternary conditions, short-circuiting logic and missing brackets!)
- full function rewrite `QueryResultsService::updateLayerResultFeatures(layer)`
- notable function rewrite `QueryResultsService::_digestFeaturesForLayers (featuresForLayers)`

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -216,245 +396,233 @@ inherit(QueryResultsService, G3WObject);
+    /**
+     * @TODO check if the first loop `features.forEach` is redundant,
+     *       it can be replaced by `_featuresToAdd.forEach` ?
+     */
+    features.forEach(feature => this._toggleLayerFeatureBox(layer, feature, true));
+    // _featuresToAdd.forEach(feature => this._toggleLayerFeatureBox(layer, feature, true));
+    _featuresToRemove.forEach(feature => this._removeLayerFeatureBox(layer, feature));
+
+    // new layer features
+    layer.features = [ ..._featuresToRemove, ..._featuresToAdd ];
+    
+    // in case of removed features
+    if (1 === _featuresToRemove.length) {
+      this._toggleLayerFeatureBox(layer, _featuresToRemove[0], false);
     }
-    //in case of removed features
-    if (findLayer.features.length === 1 && this.state.layersFeaturesBoxes[this.getBoxId(findLayer, findLayer.features[0])])
-      // used to do all vue reactive thing before update layers
-      setTimeout(() => this.state.layersFeaturesBoxes[this.getBoxId(findLayer, findLayer.features[0])].collapsed = false);
+
     // in case no more features on layer remove interaction pickcoordinate to get result from map
-    this.checkIfLayerHasNoFeatures(findLayer);
+    this.checkIfLayerHasNoFeatures(layer);
+
   }
+
   // hightlight new feature
-  this.state.layers.length === 1 && this.highlightFeaturesPermanently(this.state.layers[0]);
-  this.changeLayerResult(findLayer);
-};
+  if (1 === this.state.layers.length) {
+    this.highlightFeaturesPermanently(this.state.layers[0]);
+  }
 
-/**
- * @deprecated since v3.8. Will be deleted in 4.x. Use QueryResultsService::updateLayerResultFeatures(layer) instead
- */
-proto.addRemoveFeaturesToLayerResult = proto.updateLayerResultFeatures;
+  this.changeLayerResult(layer);
+};
```
this function has changed substantially, check it carefully

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -889,93 +1142,147 @@ proto.clearState = function(options={}) {
  * @returns {[]}
- * @private
  */
 proto._digestFeaturesForLayers = function(featuresForLayers) {
```
this is a blob function, I partially rewrote it to try to make it more readable, check it carefully

volterra79 committed:
Merge branch 'dev' into refactor-query-results-service

volterra79 committed:
Fix misspelling and properties position declaration

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -1341,99 +1697,112 @@ proto.printAtlas = function(layer, feature){
-      if (downloadsactions === undefined) action.state.toggled[index] = false;
-      else downloadsactions.state.toggled[index] = false;
-      this.setCurrentActionLayerFeatureTool({
-        index,
-        action,
-        layer
-      });
+
+    /** @FIXME add description */
+    if(features.length > 1 && undefined !== downloadsactions) {
+      layer[DownloadFormats.name].active = false;
+    }
+
+    /** @FIXME add description */
+    if (features.length <= 1 && undefined === downloadsactions) {
+      action.state.toggled[index] = false;
+    }
+
+    /** @FIXME add description */
+    if (features.length <= 1 && undefined !== downloadsactions) {
+      downloadsactions.state.toggled[index] = false;
+    }
+
+    /** @FIXME add description */
+    if (features.length <= 1) {
+      this.setCurrentActionLayerFeatureTool({ index, action, layer });
     }
   };
-  if (query.type === 'polygon'){
-    //check if multidownload if present
+
+  if ('polygon' !== query.type) {
+    runDownload();
+  } else { // check if multi-download if present
```
within this function `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`, i rewrote several if-else conditions to reduce nesting level, check carefully

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -1447,342 +1816,428 @@ proto.downloadFeatures = async function(type, layer, features=[], action, index)
-        });
-      }
+
+    /** @FIXME add description */
+    if (1 === features.length && undefined === downloadsactions) {
+      action.state.toggled[index] = true;
+    }
+
+    /** @FIXME add description */
+    if(1 === features.length) {
+      this.state.actiontools[QueryPolygonCsvAttributesComponent.name] = this.state.actiontools[layer.id] || {};
+      this.state.actiontools[QueryPolygonCsvAttributesComponent.name][layer.id] = config;
+      this.setCurrentActionLayerFeatureTool({ layer, index, action, component: QueryPolygonCsvAttributesComponent });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined === downloadsactions) {
+      layer[type].active = !layer[type].active;
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined === downloadsactions && layer[type].active) {
+        this.setLayerActionTool({ layer, component: QueryPolygonCsvAttributesComponent, config });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined === downloadsactions && !layer[type].active){
+      this.setLayerActionTool({ layer });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined !== downloadsactions) {
+      this.setLayerActionTool({ layer, component: QueryPolygonCsvAttributesComponent, config });
     }
-  } else runDownload();
+
+  }
```
within this function `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`, i rewrote several if-else conditions to reduce nesting level, check carefully

volterra79 committed:
Add comments

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -1447,342 +1816,428 @@ proto.downloadFeatures = async function(type, layer, features=[], action, index)
-    else if (hasAlreadySelectioned && force === 'remove') await layer.excludeSelectionFid(fid);
+    
+    /** @FIXME add description */
+    if (undefined === force) {
+      layer[hasAlreadySelectioned ? 'excludeSelectionFid': 'includeSelectionFid'](fid);
+    }
+
+    /** @FIXME add description */
+    if (undefined !== force && !hasAlreadySelectioned && 'add' === force) {
+      await layer.includeSelectionFid(fid);
+    }
+
+    /** @FIXME add description */
+    if (undefined !== force && hasAlreadySelectioned && 'remove' === force) {
+      await layer.excludeSelectionFid(fid);
+    }
+
+    /** @FIXME add description */
     if (layer.getFilterActive()) {
+
       const currentLayer = this.state.layers.find(_layer => _layer.id === layer.getId());
-      layer.getSelectionFids().size > 0 && currentLayer && currentLayer.features.splice(index, 1);
+
+      /** @FIXME add description */
+      if (layer.getSelectionFids().size > 0 && currentLayer) {
+        currentLayer.features.splice(index, 1);
+      }
+
       this.mapService.clearHighlightGeometry();
-      this.state.layers.length === 1 && !this.state.layers[0].features.length && this.state.layers.splice(0);
+
+      /** @FIXME add description */
+      if (1 === this.state.layers.length && !this.state.layers[0].features.length) {
+        this.state.layers.splice(0);
+      }
+
     }
+
   }
+
```
within this function `QueryResultsService::_addRemoveSelectionFeature(layer, feature, index, force)`, i rewrote several if-else conditions to reduce nesting level, check carefully

Raruto committed:
Merge branch 'dev' into refactor-query-results-service

Raruto committed:
Closes https://github.com/g3w-suite/g3w-client/issues/59

Raruto committed:
wrong parameter

Raruto committed:
add geometry check within `QueryResultsService::showGeometry(geometry)`

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -1,69 +1,205 @@
+   * @FIXME add description
+   */
   this._actions = {
     'zoomto': QueryResultsService.zoomToElement,
```
```js
QueryResultsService.zoomToElement
```

is this an undefined function?

volterra79 replied:
@Raruto yes. You can delete it. Thanks

Raruto replied:
removed in https://github.com/g3w-suite/g3w-client/pull/317/commits/40d0254fb7a5d2bf20e59990c20cde13f92694df

Raruto committed:
extract function `QueryResultsService::_handleFeatureForLayer(featuresForLayer)`

Raruto committed:
remove undefined action `QueryResultsService.zoomToElement`

Raruto committed:
clean up some code

Raruto committed:
Merge branch 'dev' into refactor-query-results-service

Raruto committed:
convert `QueryResultsService` to ES6 class

Raruto committed:
Merge branch 'dev' into refactor-query-results-service

Raruto committed:
Merge branch 'dev' into refactor-query-results-service

volterra79 committed:
Merge branch 'dev' into refactor-query-results-service

# Conflicts:
#	package.json
#	src/app/gui/queryresults/queryresultsservice.js

Raruto committed:
update @since tags

Raruto committed:
update @param tag

volterra79 committed:
Merge branch 'refactor-query-results-service' of https://github.com/g3w-suite/g3w-client into refactor-query-results-service

# Conflicts:
#	src/app/gui/queryresults/queryresultsservice.js

volterra79 committed:
:bug: Fix bugs due refactoring

Raruto committed:
wrong merge

volterra79 committed:
:bug: Fix undefined check and change layer.selection.features to empty array instead empty object

Raruto committed:
code readability

- slim down function  `updateLayerResultFeatures()`
- alias long variable name `this._addFeaturesLayerResultInteraction`
- slim down function `downloadFeatures()`
- extract function `_handleExternalVectorLayerSelection()`
- extract function `_handleProjectLayerSelection()`

volterra79 committed:
Merge branch 'dev' into refactor-query-results-service

volterra79 committed:
:bug: Ensure that find array method return undefined (not found). Add extenal parameter into _getFeaturesIds

Raruto committed:
slim down `unRegisterCustomComponent`

volterra79 committed:
:bug: Fix add feature to a external layer. fix add settimeout to querycontrol

Raruto committed:
code readability

- slim down comments
- alias long variable name `this.state.currentactiontools`
- alias long variable name `this.state.currentactionfeaturelayer`
- alias long variable name ` this.state.layersFeaturesBoxes`
- extract function `_clearActions`
- alias prototype function `QueryResultsService.init`
- alias prototype function `QueryResultsService.reset`

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -1,1874 +1,2555 @@
+        this.setLayerActionTool({ layer });
+      }
+
+      const download_caller_id = ApplicationService.setDownload(true);
 
+      GUI.setLoadingContent(true);
+
+      try {
+        await CatalogLayersStoresRegistry.getLayerById(layer.id).getDownloadFilefromDownloadDataType(type, { data }) || Promise.resolve();
+      } catch(err) {
+        GUI.notify.error(err || t("info.server_error"));
+      }
+
+      ApplicationService.setDownload(false, download_caller_id);
+
+      GUI.setLoadingContent(false);
```
@volterra79 after, try to see if you are able to rewrite these kind of statements to make use of some sort of download wrapper (eg. using some other `utils` function I'm not aware of):

```suggestion
      await this.downloadApplicationWrapper(
        CatalogLayersStoresRegistry
          .getLayerById(layer.id)
          .getDownloadFilefromDownloadDataType
          .bind(this, type, { data })
      );
```

Raruto committed:
refactor function `QueryControl.setMap()`

Raruto committed:
refactor `_handleFeatureForLayer`

volterra79 committed:
:recycle: Set right event 'picked' on queryControl

Raruto committed:
comments

volterra79 committed:
:bug: Fix setCurrentActionLayerFeatureTool

volterra79 committed:
Check and correct some refactoring code

volterra79 committed:
:sparkles: Create common downloadApplicationWrapper to run download

volterra79 committed:
:recycle: Move downloadApplicationWrapper from queryresultsservice.js to index.prod.js as GUI method

Raruto committed:
move application wrapper

Raruto committed:
extract `_hasLayerObjGeometry` and `_parseLayerObjFeatures`

Raruto committed:
params destrcuturing: `featuresForLayer`

Raruto committed:
simplify condition

Raruto committed:
refactor `setQueryResponse`

- remove function
 `_digestFeaturesForLayers()`
- rename function
 `_handleFeatureForLayer()` into `_responseToLayer()`
- extract function `_hasLayerObjGeometry()`
- extract function `_hasLayerObjImageField()`
- refactor function `_parseLayerObjFormStructure()`

Raruto committed:
alias `_changeLayerResult`

Raruto committed:
remove `_orderResponseByProjectLayers`

Raruto committed:
apply some fixme and todos

- remove nested if-else in: `getVectorLayerFeaturesFromQueryRequest()`
- incorporate function `_addVectorLayersDataToQueryResponse()` into `*/
  setQueryResponse()` setter
- simplify destructuring assignment within `_printSingleAtlas()` function
- remove function `_addComponent()`

Raruto committed:
Merge branch 'dev' into refactor-query-results-service

volterra79 committed:
Merge branch 'dev' into refactor-query-results-service

Raruto committed:
wrong variable reference

Raruto committed:
Merge branch 'refactor-query-results-service' of https://github.com/g3w-suite/g3w-client into refactor-query-results-service

Raruto committed:
wrong variable name

volterra79 committed:
:bug: Fix case when click to add feature to layer result and no current mapcontrol is active (undefined). Need to be set null to align default value and evaluation.

Raruto committed:
`console.warn`

Raruto committed:
wrong comments

volterra79 committed:
:bug: In case of no features for a specific layer, return undefind, so layer is not add to query results state layers array

Raruto committed:
simplify conditions

volterra79 committed:
:bug: Need to refer to instance and not to "class"

volterra79 committed:
:bug: Handle selection feature action when add new feature to layer result

volterra79 committed:
:bug: Add custom drawpolygon type for query by drw polygon otherwise it conflict to download action to get all attributes of polygon query (query by polygon)

volterra79 committed:
:bug: fix download action

Raruto committed:
boolean logic

volterra79 committed:
Merge branch 'dev' into refactor-query-results-service
